L44-python-rich-status.md

```diff
@@ -0,0 +1,474 @@
+
+### Behavior Different Across Implementations
+
+C++ implementation tolerates anything to be put in the `grpc-status-details-bin` entry; Java checks if the code of the proto status is matching the status code of the RPC call; Golang enforces both code and message of the proto status, and the RPC call be the same.
+
+So, which paradigm should Python follow?
+
+### The Proto of Status
+
+The proto of status is well-defined and used in many frameworks. Here is the definition of `Status`, for full version see [status.proto](https://github.com/googleapis/api-common-protos/blob/master/google/rpc/status.proto).
+```proto
+message Status {
+  int32 code = 1;
+  string message = 2;
+  repeated google.protobuf.Any details = 3;
+}
+```
+
+### Current Python API About RPC Status
+
+```Python
+# Client side
+stub = ...Stub(channel)
+try:
+    resp = stub.ARpcCall(...)
+except grpc.RpcError as rpc_error:
+    code = rpc_error.code()
+    details = rpc_error.details()
+    # Unable to get rich status
+```
+
+```Python
+# Server side
+def ...Servicer(...):
+
+    def ARPCCall(request, context):
+        context.set_code(...)
+        context.set_details(...)
+        # Unable to set rich status
+
+```
+
+## Proposal
+
+### A New Public Interface: Status
+
+This class is used to describe the status of an RPC.
+
+The name of the class is `Status`, because it is shorter and cleaner than `RpcStatus` or `ServerRpcStatus`. In the future, we might want to utilize it at client side as well.
+
+The metadata field of `Status` is `trailing_metadata` instead of `metadata`. The `Status` should be the final status of an RPC, so it should assist developers to get information that only accessible at the end of the RPC. Also, if we want to support it at client side, we need to separate `initial_metadata` and `trailing_metadata`.
+
+```Python
+class Status(six.with_metaclass(abc.ABCMeta)):
+    """Describes the status of an RPC.
+
+    This is an EXPERIMENTAL API.
+
+    Attributes:
+      code: A StatusCode object to be sent to the client.
+      details: An ASCII-encodable string to be sent to the client upon
+        termination of the RPC.
+      trailing_metadata: The trailing :term:`metadata` in the RPC.
+    """
+```
+
+### Server side API
+
+gRPC Python team encountered lots of disagreements about how this part should be implemented. Available options are listed in the `Rationale` section. Here is our final consensus as a team, but feel free to comment about other options.
```

Comment:
I suggest removing the first sentence about internal disagreements - I don't think it adds anything for future readers...

Reply:
Removed.